tests/cpp/test_sharding.cpp

```diff
@@ -262,13 +262,102 @@ TEST_F(ShardingTest, ResidualAdd) {
 
   preseg_passes::OptimizationPass<
       preseg_passes::PropagateShardingsPass>::runPass(fusion.get());
+  // getShardedLogicalAxis uses maybeAllocationDomain to get the sharded axis.
+  // Setting allocation domain here manually, which is otherwise done by
+  // MakeReshardingContiguousPass, to isolate the test to a single preseg pass.
+  for (auto tv : fusion->allTvs()) {
+    tv->setAllocationDomain(tv->getLoopDomain(), true);
+  }
   NVF_CHECK(tv1->hasDeviceMesh());
+  int64_t expected_sharded_axis =
+      getShardedLogicalAxis(tv0, ParallelType::DIDx);
+  NVF_CHECK(expected_sharded_axis != -1, "tv0 should have a sharded axis.");
   NVF_CHECK(
-      getShardedLogicalAxis(tv1, ParallelType::DIDx) ==
-          getShardedLogicalAxis(tv0, ParallelType::DIDx),
+      getShardedLogicalAxis(tv1, ParallelType::DIDx) == expected_sharded_axis,
       "Expected tv1 to be sharded like tv0 due to backpropagation of shardings.");
 }
 
+TEST_F(ShardingTest, PropagateParallelTypeOnce) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  DeviceMesh mesh({0, 1});
+
+  TensorView* tv0 = makeContigTensor(2);
+  TensorView* tv1 = makeContigTensor(2);
+  TensorView* tv2 = add(tv0, tv1);
+
+  tv0->setDeviceMesh(mesh);
+  tv0->outer_split(0, mesh.size());
+  tv0->axis(0)->parallelize(ParallelType::DIDx);
+
+  tv1->setDeviceMesh(mesh);
+  tv1->outer_split(1, mesh.size());
+  tv1->axis(1)->parallelize(ParallelType::DIDx);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addOutput(tv2);
+
+  preseg_passes::OptimizationPass<
+      preseg_passes::PropagateShardingsPass>::runPass(fusion.get());
+  for (auto tv : fusion->allTvs()) {
+    tv->setAllocationDomain(tv->getLoopDomain(), true);
+  }
+  NVF_CHECK(numDeviceDims(tv2) == 1);
+  int64_t expected_sharded_axis =
+      getShardedLogicalAxis(tv0, ParallelType::DIDx);
```

Comment:
Note: `getShardedLogicalAxis` cannot use loop domain currently since it is called for tvs after scheduling and may have merge transforms not supported by this utility. I will think about potential solutions which may allow us to use loop domain here.

Reply:
> getShardedLogicalAxis cannot use loop domain currently

It doesn't and IMO shouldn't. getShardedLogicalAxis is used to connect TVs with concrete tensors. That connection should respect allocation domain not loop domain. Is my understanding correct?

Reply:
Yes, that makes sense.